Make HTTP_ a constants to avoid creating the string every loop

martinemde opened the pull request:
I noticed we were getting a lot of strings allocated from utf8-sanitizer during requests, so I tooke the opportunity to reduce the string allocation.

See [my gist with memory_profiler output for the test run before and after](https://gist.github.com/martinemde/db2e08db793621d4d1da/revisions). The gist has a few revisions which show the difference between the two runs as a diff instead of trying to compare outputs.

```
-Total allocated: 143837 bytes (1955 objects) # before
+Total allocated: 132357 bytes (1668 objects) # after
```

bf4 commented on lib/rack/utf8_sanitizer.rb:
```diff
@@ -61,17 +68,17 @@ def sanitize_rack_input(env)
-      content_type   = env['CONTENT_TYPE']
-      content_type &&= content_type.split(/\s*[;,]\s*/, 2).first
+      content_type   = env[CONTENT_TYPE]
+      content_type &&= content_type.split(CONTENT_TYPE_SPLIT, 2).first
```
I think regexps are always cached, or at least we can set the `o` flag when there's no interpolation?

martinemde replied:
That may be true. I wasn't as sure about adding the regexp constants and I'd be fine pulling them out or trying a profiler run with and without.

bf4 commented:
@martinemde Thanks for this. What was the command/config you ran?  The gist doesn't include it. I could guess, but easier to ask :)

martinemde commented:
On my phone right now, but I just added memory_profiler gem, wrapped the main describe in the tests with memory_profiler as described in the gem docs, and added `(allow_files: "lib/rack/utf8_sanitizer")`.

Start after the requires and end at the end of the test file, that's it.

martinemde commented:
Btw, the noticeable ones in our application were "HTTP_" and "C".

martinemde commented on lib/rack/utf8_sanitizer.rb:
```diff
@@ -155,8 +163,8 @@ def decode_string(input)
-      input.gsub(/%([a-f\d]{2})/i) do |encoded|
-        decoded = [$1.hex].pack('C')
+      input.gsub(PERCENT_CHAR_REGEXP) do |encoded|
+        decoded = [$1.hex].pack(PACK_C)
```
I just noticed the array here too. It might be worth benchmarking `$1.hex.char` vs this?

bf4 replied:
I know `$1.hex.chr` can blow up.. I don't think `pack('C')` would.

```ruby
"\u{1F4A9}" # => "💩"
"1F4A9".hex.chr #=> RangeError: 128169 out of char range
["1F4A9".hex].pack('C') # => "\xA9"
```

whitequark replied:
/%([a-f\d]{2})/i

There's no way `$1.hex.chr` blows up with two nibbles.

martinemde replied:
I showed that it won't, as you say, in PR #36.

bf4 replied:
Good call. I overlooked the `{2}`

whitequark commented:
Can we use the frozen string optimization from Ruby 2.3 instead of this? Constant lookups aren't free either.

martinemde commented:
Ok, interestingly, the only 2 strings that matter in this file are `HTTP_` and the `C` I already removed in #36. Check out the [new results](https://gist.github.com/martinemde/abdda997264b8e951dc8/revisions). I notice also that `CONTENT_LENGTH` gets allocated once per iteration, but not `CONTENT_TYPE` and I have no idea why that is.

whitequark commented on lib/rack/utf8_sanitizer.rb:
```diff
@@ -67,7 +69,7 @@ def sanitize_rack_input(env)
       uri_encoded = URI_ENCODED_CONTENT_TYPES.any? {|type| content_type == type}
 
-      if env["rack.input"]
+      if env['rack.input']
```
I don't think this change makes any difference?

martinemde replied:
yep, that was my mistake when changing it back to strings. fixed.

martinemde commented:
This is pretty simple now... just changes HTTP_ to a constant to avoid making it each time we loop through the headers.

martinemde commented:
I updated my commit to reflect what this actually is now.
